File: Python/Code.py
def rabin_karp(text, pattern):
    text_length = len(text)
    pattern_length = len(pattern)
    prime = 101  # Prime number for hashing

    # Calculate the hash value for the pattern and the first window of the text
    pattern_hash = sum(ord(pattern[i]) * pow(prime, i) for i in range(pattern_length))
    text_hash = sum(ord(text[i]) * pow(prime, i) for i in range(pattern_length))

    # Iterate through the text to find the pattern
    for i in range(text_length - pattern_length + 1):
        if pattern_hash == text_hash and text[i:i + pattern_length] == pattern:
            return i  # Pattern found at index i

        # Update the hash value for the next window
        if i < text_length - pattern_length:
            text_hash = (text_hash - ord(text[i])) / prime
            text_hash = (text_hash * prime + ord(text[i + pattern_length])) % prime

    return -1  # Pattern not found in text

text = "abcdefghijk"
pattern = "abc"

index = rabin_karp(text, pattern)
if index != -1:
    print(f"Pattern '{pattern}' found in text '{text}' at index {index}")
else:
    print(f"Pattern '{pattern}' not found in text '{text}'")

File: C++/Code.cpp
#include <iostream>
#include <string>

using namespace std;

int rabin_karp(string text, string pattern, int prime) {
    int text_length = text.length();
    int pattern_length = pattern.length();
    int i, j;
    int p = 0;
    int t = 0;
    int h = 1;

    for (i = 0; i < pattern_length - 1; i++)
        h = (h * 256) % prime;

    for (i = 0; i < pattern_length; i++) {
        p = (256 * p + pattern[i]) % prime;
        t = (256 * t + text[i]) % prime;
    }

    for (i = 0; i <= text_length - pattern_length; i++) {
        if (p == t) {
            for (j = 0; j < pattern_length; j++) {
                if (text[i + j] != pattern[j])
                    break;
            }

            if (j == pattern_length)
                return i;
        }

        if (i < text_length - pattern_length) {
            t = (256 * (t - text[i] * h) + text[i + pattern_length]) % prime;

            if (t < 0)
                t += prime;
        }
    }

    return -1;
}

int main() {
    string text = "abcdefghijkhrt";
    string pattern = "hrt";
    int prime = 101;

    cout << "Text: " << text << endl;
    cout << "Pattern: " << pattern << endl;

    int index = rabin_karp(text, pattern, prime);

    if (index != -1)
        cout << "Pattern found at index: " << index << endl;
    else
        cout << "Pattern not found" << endl;

    return 0;
}
